src/storage/src/source/sql_server/replication.rs

```diff
@@ -0,0 +1,337 @@
+                            cdc_handle.start_lsn(&output_info.capture_instance, initial_lsn);
+                    }
+                    _ => unreachable!(""),
+                }
+            }
+
+            // Off to the races! Replicate data from SQL Server.
+            let cdc_stream = cdc_handle
+                .poll_interval(std::time::Duration::from_millis(500))
+                .into_stream();
+            let mut cdc_stream = std::pin::pin!(cdc_stream);
+
+            while let Some(event) = cdc_stream.next().await {
+                let event = event.map_err(TransientError::from)?;
+                tracing::debug!(?event, "got replication event");
+
+                let (capture_instance, commit_lsn, changes) = match event {
+                    // We've received all of the changes up-to this LSN, so
+                    // downgrade our capability.
+                    CdcEvent::Progress { next_lsn } => {
+                        tracing::debug!(?next_lsn, "got a closed lsn");
+                        data_cap_set.downgrade(Antichain::from_elem(next_lsn));
+                        upper_cap_set.downgrade(Antichain::from_elem(next_lsn));
+                        continue;
+                    }
+                    // We've got new data! Let's process it.
+                    CdcEvent::Data {
+                        capture_instance,
+                        lsn,
+                        changes,
+                    } => (capture_instance, lsn, changes),
+                };
+
+                // Decode the SQL Server row into an MZ one.
+                let Some((partition_idx, decoder)) = capture_instances.get(&capture_instance)
+                else {
+                    let definite_error = DefiniteError::ProgrammingError(format!(
+                        "capture instance didn't exist: '{capture_instance}'"
+                    ));
+                    let () = return_definite_error(
+                        definite_error,
+                        &output_indexes[..],
+                        data_output,
+                        data_cap_set,
+                        definite_error_handle,
+                        definite_error_cap_set,
+                    )
+                    .await;
+                    return Ok(());
+                };
+                let data_cap = data_cap_set.delayed(&commit_lsn);
+
+                for change in changes {
+                    let (sql_server_row, diff): (_, _) = match change {
+                        CdcOperation::Insert(sql_server_row)
+                        | CdcOperation::UpdateNew(sql_server_row) => (sql_server_row, Diff::ONE),
+                        CdcOperation::Delete(sql_server_row)
+                        | CdcOperation::UpdateOld(sql_server_row) => {
+                            (sql_server_row, Diff::MINUS_ONE)
+                        }
+                    };
+
+                    // Failing to decode data is a permanent failure.
```

Comment:
Same as above, we should continue without halting here

Reply:
Done!